libtelemetry/src/main/java/com/mapbox/android/telemetry/CertificateBlacklist.java

```diff
@@ -0,0 +1,202 @@
+      try {
+        List<String> blacklist = obtainBlacklistContents(file);
+        lastUpdateTime = Long.valueOf(blacklist.get(BLACKLIST_HEAD));
+      } catch (IOException exception) {
+        Log.e(LOG_TAG, RETRIEVE_TIME_FAIL, exception);
+      }
+    }
+
+    return lastUpdateTime;
+  }
+
+  void updateBlacklist() {
+    Request request = new Request.Builder()
+      .url(determineConfigEndpoint() + QUESTION_MARK + ACCESS_TOKEN_QUERY_PARAMETER + EQAL_SIGN + accessToken)
+      .build();
+
+    OkHttpClient client = new OkHttpClient();
+    client.newCall(request).enqueue(this);
+  }
+
+  private void saveBlackList(List<String> revokedKeys) {
+    String fileContents = createListContent(revokedKeys);
+    FileOutputStream outputStream;
+
+    try {
+      outputStream = context.openFileOutput(BLACKLIST_FILE, Context.MODE_PRIVATE);
+      outputStream.write(fileContents.getBytes());
+      outputStream.close();
+    } catch (Exception exception) {
+      Log.e(LOG_TAG, SAVE_BLACKLIST_FAIL, exception);
+    }
+  }
+
+  @Override
+  public void onFailure(Call call, IOException exception) {
+    Log.e(LOG_TAG, REQUEST_FAIL, exception);
+  }
+
+  @Override
+  public void onResponse(Call call, Response response) throws IOException {
+    List<String> revokedKeys = extractResponse(response);
+
+    saveBlackList(revokedKeys);
+  }
+
+  private String createListContent(List<String> revokedKeys) {
+    Date date = new Date();
+
+    StringBuilder content = new StringBuilder(date.getTime() + NEW_LINE);
+
+    for (String key: revokedKeys) {
+      content.append(SHA256).append(key).append(NEW_LINE);
+    }
+
+    return content.toString();
+  }
+
+  private List<String> obtainBlacklistContents(File file) throws IOException {
+    InputStream inputStream = new FileInputStream(file);
+    BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
+
+    List<String> blacklist = new ArrayList<>();
+
+    boolean done = false;
+
+    while (!done) {
+      String line = reader.readLine();
+      done = (line == null);
+
+      if (line != null && !line.isEmpty()) {
+        blacklist.add(line);
+      }
+    }
+
+    reader.close();
+    inputStream.close();
+
+    return blacklist;
+  }
+
+  private String determineConfigEndpoint() {
+    EnvironmentChain environmentChain = new EnvironmentChain();
+    EnvironmentResolver setupChain = environmentChain.setup();
+    ServerInformation serverInformation;
+    try {
+      ApplicationInfo appInformation = context.getPackageManager().getApplicationInfo(context.getPackageName(),
+        PackageManager.GET_META_DATA);
+
+      if (appInformation != null && appInformation.metaData != null) {
+        serverInformation = setupChain.obtainServerInformation(appInformation.metaData);
+        return ENDPOINTS.get(serverInformation.getEnvironment());
+      }
+    } catch (Exception exception) {
+      Log.e(LOG_TAG, ENDPOINT_DETERMINATION_FAIL, exception);
+    }
+
+    return null;
```

Comment:
`context.getPackageManager().getApplicationInfo` could throw a `PackageManager.NameNotFoundException` - this shouldn't happen but if it does we'd return `null` which would cause a crash in the call site https://github.com/mapbox/mapbox-events-android/blob/f4bba88eda2949f5d1b6d88592ca2e290efc6be8/libtelemetry/src/main/java/com/mapbox/android/telemetry/CertificateBlacklist.java#L108